ledger/mruaccts.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"container/list"
+
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/logging"
+)
+
+// mruAccounts provides a storage class for the most recently used accounts data.
+// It doesn't have any syncronization primitive on it's own and require to be
+// syncronized by the caller.
+type mruAccounts struct {
+	// accountsList contain the list of persistedAccountData, where the front ones are the most "fresh"
+	// and the ones on the back are the oldest.
+	accountsList *list.List
+	// accounts provides fast access to the various elements in the list by using the account address
+	accounts map[basics.Address]*list.Element
+	// pendingAccounts are used as a way to avoid taking a write-lock. When the caller needs to "materialize" these,
+	// it would call flushPendingWrites and these would be merged into the accounts/accountsList
+	pendingAccounts chan persistedAccountData
+	// log interface; used for logging the threshold event.
+	log logging.Logger
+	// pendingWritesWarnThreshold is the threshold beyond we would write a warning for exceeding the number of pendingAccounts entries
+	pendingWritesWarnThreshold int
+}
+
+// init initializes the mruAccounts for use.
+// thread locking semantics : write lock
+func (m *mruAccounts) init(log logging.Logger, pendingWrites int, pendingWritesWarnThreshold int) {
```

Comment:
The `New` is meant for allocating a new object and returning a pointer-to-the-object, whereas here, I just wanted to initialize already existing object. I tried to avoid the "New" pattern here since I didn't want to handle the nil baseAccount use case.

As for the sub package idea - I don't think that it would be appropriate for the mruAccounts alone to be in a sub-package. I think that we need the entire accountUpdates to be moved into it's own sub package. Let's do that in a separate PR ?